app/async/Worker.php

```diff
@@ -0,0 +1,236 @@
+    {
+        $this->asyncJobs->beginTransaction();
+        try {
+            // get jobs in the actual order they should be executed
+            $jobs = $this->asyncJobs->findAllReadyForExecution($this->pollingInterval, $workerId);
+            if (!$jobs) {
+                $this->asyncJobs->rollback();
+                return null;
+            }
+
+            $selectedJob = reset($jobs); // default is the first job returned from repository
+            foreach ($jobs as $job) {
+                if ($job->getScheduledAt() && $this->getJobDispatchDelay($job) === 0) {
+                    // but if we have a scheduled job that already should have been executed...
+                    $selectedJob = $job; // ... we make it a priority!
+                    break;
+                }
+            }
+
+            $selectedJob->allocateForWorker($workerId);
+            $this->asyncJobs->persist($selectedJob);
+            $this->asyncJobs->commit();
+        } catch (Exception $e) {
+            $this->asyncJobs->rollback();
+            throw $e;
+        }
+        return $selectedJob;
+    }
+
+    /**
+     * Computes after how many seconds a job should be executed.
+     * @return int number of seconds (0 = immediately)
+     */
+    private function getJobDispatchDelay(AsyncJob $job): int
+    {
+        if ($job->getScheduledAt()) {
+            $delay = $job->getScheduledAt()->getTimestamp() - time();
+            return max($delay, 0); // past jobs are represented as immediate jobs
+        } else {
+            return 0; // other jobs are scheduled immediately
+        }
+    }
+
+    /**
+     * Process an actual async job.
+     * @param AsyncJob $job to be processed
+     */
+    private function dispatchJob(AsyncJob $job)
+    {
+        if ($job->getRetries() > $this->retries) {
+            // this job has been beaten to death...
+            $job->setTerminatedNow();
+            $this->asyncJobs->persist($job);
+            $this->asyncJobs->flush();
+            return;
+        }
+
+        --$this->jobsRemaining; // this counter goes down even in case of failure
+
+        try {
+            $this->dispatcher->dispatch($job);
+            $job->setTerminatedNow();
+        } catch (Exception $e) {
+            $job->appendError($e->getMessage());
+            $this->logger->log(
+                sprintf("Job '%s' failed (try #%d): %s", $job->getCommand(), $job->getRetries(), $e->getMessage()),
+                ILogger::ERROR
+            );
+        }
+
+        $this->asyncJobs->persist($job);
+        $this->asyncJobs->flush();
+    }
+
+    /**
+     * Main method of the worker that actually processes the async job.
+     * This method is blockinig, once it terminates the worker process should also terminate.
+     * @param string $workerId identifier of the worker (should be gathered from the CLI arguments)
+     */
+    public function run(string $workerId = 'worker1')
```

Comment:
I don't think that this should be parameter with default value, it should definitely be provided by caller everytime, to avoid future mistakes.